### PR TITLE
Change sharing criterion to avoid artificial detours

### PR DESCRIPTION
**Pull request by genadz**
Changes:
- use continuous function to calculate sharing coefficient (instead of piecewise constant function);
- change definition of "sharing": it was modified  from "part of edges that a candidate shares" to "part of edges that already chosen path shares with alternative";
- extend alternatives search.

## Tasklist

 - [ ] Add tests
 - [ ] Add #fixes with the issue number that this PR addresses
 - [ ] Update the docs with any new request parameters or changes to behavior described
 - [ ] Update the [changelog](CHANGELOG.md)
 - [ ] If you made changes to the lua files, update the [taginfo](taginfo.json) too.

## Requirements / Relations

 Link any requirements here. Other pull requests this PR is based on?

### CODE REVIEW

**Comment by genadz** [comment]
## Quantity

### ETA < 10m

| | master | branch |
| --- | --- | --- |
| 1st alternative | 1454 | 1463 |
| 2nd alternative | 918 | 930 |

### 10m < ETA < 30m

| | master | branch |
| --- | --- | --- |
| 1st alternative | 511 | 508 |
| 2nd alternative | 356 | 352 |

### 30m < ETA < 1h

| | master | branch |
| --- | --- | --- |
| 1st alternative | 707 | 754 **(+7%)** |
| 2nd alternative | 545 | 602 **(+10%)** |

### 1h < ETA < 5h

| | master | branch |
| --- | --- | --- |
| 1st alternative | 3537 | 4045 **(+14%)** |
| 2nd alternative | 2360 | 3201 **(+36%)** |

## Quality

Let me share several examples:

### with alternatives only in master

| before | after |
| --- | --- |
| master_1 | branch_1 |
| master_2 | branch_2 |
| master_3 | branch_3 |

As expected, some alternatives that have unreasonable detours were filtered out.

### with alternatives only in branch

| before | after |
| --- | --- |
| master_4 | branch_4 |
| master_1 | branch_1 |

These changes caused by increased `kAlternativeCostExtend` parameter.

### different alternatives were proposed

| before | after |
| --- | --- |
| master_1 | branch_1 |
| master_3 | branch_3 |
| master_2 | branch_2 |

File: src/thor/alternates.cc [hunk 1]
```diff
@@ -28,24 +28,18 @@ namespace thor {
...
-  // 20km
-  else if (distance < 20000.) {
-    return 0.60;
-  }
-  // 50km
-  else if (distance < 50000.) {
-    return 0.65;
-  }
+  if (distance < 10000.f)
+    return 0.6f;
```
Comment (genadz) [inline src/thor/alternates.cc:35]: changed min `at_most_shared` coefficient from `0.5` to `0.6`

File: src/thor/alternates.cc [hunk 2]
```diff
@@ -28,24 +28,18 @@ namespace thor {
...
   // 100km
-  else if (distance < 100000.) {
-    return 0.70;
+  if (distance < 100000.f) {
+    // Uniformly increase 'at_most_shared' value from 0.6 to 0.75 for routes
+    // from 10km to 100km
+    return 0.6f + (kAtMostShared - 0.6f) * (distance - 10000.f) / (100000.f - 10000.f);
```
Comment (genadz) [inline src/thor/alternates.cc:40]: replaced piecewise constant function with continuous one

File: src/thor/alternates.cc [hunk 3]
```diff
@@ -175,20 +169,19 @@ bool validate_alternate_by_sharing(std::vector<std::unordered_set<GraphId>>& sha
...
         shared_length += length;
       }
     }
 
-    // throw this alternate away if it shares more than at_most_shared with any of the chosen paths
-    assert(total_length > 0);
-    if ((shared_length / total_length) > at_most_shared) {
+    // throw this alternate away if any of the chosen paths shares more than at_most_shared with it
+    assert(paths[i].back().path_distance > 0);
+    if ((shared_length / paths[i].back().path_distance) > at_most_shared) {
```
Comment (genadz) [inline src/thor/alternates.cc:184]: this is  important change. What was before:
"sharing value is a part of shared edges of an alternative candidate with some already chosen route"
now is
"sharing value is a part of shared edges of some already chosen route with an alternative candidate".

File: src/thor/bidirectional_astar.cc [hunk 4]
```diff
@@ -28,7 +28,7 @@ constexpr float kThresholdDelta = 420.0f;
 // an upper bound value cost for alternative routes we're looking for. Due to the fact that
 // we can't estimate route cost that goes through some particular edge very precisely, we
 // can find alternatives with costs greater than the threshold.
-constexpr float kAlternativeCostExtend = 1.1f;
+constexpr float kAlternativeCostExtend = 1.2f;
```
Comment (genadz) [inline src/thor/bidirectional_astar.cc:31]: It doesn't drop performance too much, I will add numbers later

File: src/thor/alternates.cc (hunk 3 above)
Comment (genadz) [reply src/thor/alternates.cc:184]: Let's consider a particular example:
```
A------------------------------------B---------C----------------------------------------D
                                     |         |
                                     |         |
                                     |         |
                                     |         |
                                     E---------F
```
here `ABCD` is an optimal route and `ABEFCD` is an alternative candidate.
Let's assume that `len(AB + CD) / len(ABCD) = 0.85` and `len(AB + CD) / len(ABEFCD) = 0.75`. Current **master** will allow `ABEFCD` to be an alternative despite the fact that `BEFC` detour is obviously unreasonable. This **branch** will reject this alternative due to the fact that `sharing = 0.85 > 0.75 = at_most_shared` value.

Comment (kevinkreiser) [reply src/thor/alternates.cc:184]: this was what was originally intended actually... good catch! i should also note though that in another pr we are looking into problems with teh path_distance not being set on the last edge label, which is where the pathinfo gets it. it is my h unch that this is in the case that we find a path using only the forward search tree but i havent proven it yet.

**Review by kevinkreiser — DISMISSED** [review]

File: src/thor/alternates.cc (hunk 3 above)
Comment (genadz) [reply src/thor/alternates.cc:184]: Hm, thanks for the comment,  I will check it

Comment (genadz) [reply src/thor/alternates.cc]: @kevinkreiser replaced division with multiplication to avoid possible problems

**Comment by merkispavel** [comment]
I guess this is a sign that valhalla's Gods like your changes 
image

**Review by merkispavel — APPROVED** [review]
reapprove, looks good

**Pull request merged**